packages/application-config/src/substitute-env-variable-placeholders.ts

```diff
@@ -0,0 +1,69 @@
+/**
+ * NOTE:
+ *  Allows env variable placeholders e.g. `${env:MC_API_URL}`.
+ *  Other placeholder types might be supported in the future.
+ */
+const variableSyntax = /\${([ ~:a-zA-Z0-9._\'",\-\/\(\)]+?)}/g;
+const envRefSyntax = /^env:/g;
+
+const hasVariablePlaceholder = (valueOfEnvConfig: string) =>
+  typeof valueOfEnvConfig === 'string' &&
+  // Using `{regex}.test()` might cause false positives if called multiple
+  // times on a global regular expression:
+  // See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/test
+  //    As with exec() (or in combination with it), test() called multiple times
+  //    on the same global regular expression instance will advance past the previous match.
+  Boolean(valueOfEnvConfig.match(variableSyntax));
+
+const isEnvVariablePlaceholder = (valueOfPlaceholder: string) =>
+  Boolean(valueOfPlaceholder.match(envRefSyntax));
+
+const substituteEnvVariablePlaceholder = (
+  valueOfPlaceholder: string,
+  matchedString: string,
+  valueOfEnvConfig: string
+) => {
+  const [, requestedEnvVar] = valueOfPlaceholder.split(':');
+  const valueOfEnv = process.env[requestedEnvVar];
+
+  if (!valueOfEnv) {
+    throw new Error(
+      `Missing '${requestedEnvVar}' specified in config as 'env:${requestedEnvVar}'.`
+    );
+  }
+
+  const escapedMatchedString = matchedString.replace(/[${}:]/g, '\\$&');
+  return valueOfEnvConfig.replace(
+    new RegExp(`(${escapedMatchedString})+`, 'g'),
+    valueOfEnv
+  );
+};
+
+const getValueOfPlaceholder = (valueWithPlaceholder: string) =>
+  valueWithPlaceholder
+    .replace(variableSyntax, (_match, varName) => varName.trim())
+    .replace(/\s/g, '');
+
+const substituteEnvVariablePlaceholders = <T>(config: T): T =>
+  JSON.parse(JSON.stringify(config), (_key, value) => {
```

Comment:
FYI: Using JSON.parse and the callback function simplifies the nested recursion logic.